### PR TITLE
Fix/compute unqualified nodes

### PR DESCRIPTION
**Pull request by sasurobert**
fix compute unqualified nodes when there are no nodes to unstake - it can happen when someone unstake tokens but all of its nodes are in jail
optimize get total topup and total stake
fix: set totUp if it is less than 0 to 0

### CODE REVIEW

**Review by iulianpascalau — DISMISSED** [review]

**Review by raduchis — DISMISSED** [review]

File: epochStart/metachain/stakingDataProvider.go [hunk 1]
```diff
@@ -268,20 +266,21 @@ func (sdp *stakingDataProvider) getValidatorInfoFromSC(validatorAddress string)
...
 	}
 	if vmOutput.ReturnCode != vmcommon.Ok {
-		return nil, nil, nil, fmt.Errorf("%w, error: %v", epochStart.ErrExecutingSystemScCode, vmOutput.ReturnCode)
+		return nil, nil, nil, nil, fmt.Errorf("%w, error: %v", epochStart.ErrExecutingSystemScCode, vmOutput.ReturnCode)
 	}
 
-	if len(vmOutput.ReturnData) < 2 {
-		return nil, nil, nil, fmt.Errorf("%w, getTotalStakedTopUpBlsKeys function should have at least two values", epochStart.ErrExecutingSystemScCode)
+	if len(vmOutput.ReturnData) < 3 {
+		return nil, nil, nil, nil, fmt.Errorf("%w, getTotalStakedTopUpBlsKeys function should have at least two values", epochStart.ErrExecutingSystemScCode)
```
Comment (SebastianMarian) [inline epochStart/metachain/stakingDataProvider.go]: at least three values

File: epochStart/metachain/stakingDataProvider_test.go [hunk 2]
```diff
@@ -121,7 +121,7 @@ func TestStakingDataProvider_PrepareDataForBlsKeyLoadOwnerDataErrorsShouldErr(t
 	err = sdp.loadDataForBlsKey([]byte("bls key"))
 	assert.NotNil(t, err)
 	assert.True(t, strings.Contains(err.Error(), epochStart.ErrExecutingSystemScCode.Error()))
-	assert.True(t, strings.Contains(err.Error(), "getTotalStakedTopUpBlsKeys function should have at least two values"))
```
Comment (iulianpascalau) [inline epochStart/metachain/stakingDataProvider_test.go:137]: three

**Review by SebastianMarian — DISMISSED** [review]

**Review by raduchis — DISMISSED** [review]

**Review by dorin-iancu — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Review by raduchis — DISMISSED** [review]

**Review by dorin-iancu — APPROVED** [review]

File: vm/systemSmartContracts/staking.go [hunk 3]
```diff
@@ -565,12 +565,15 @@ func (s *stakingSC) unStakeAtEndOfEpoch(args *vmcommon.ContractCallInput) vmcomm
 		return vmcommon.Ok
 	}
 
-	if !registrationData.Staked {
-		log.Debug("stakingSC.unStakeAtEndOfEpoch: cannot unStake node which was already unStaked")
+	if !registrationData.Staked && !registrationData.Waiting {
```
Comment (raduchis) [inline vm/systemSmartContracts/staking.go:568]: the method unStakeAtEndOfEpoch has some similarities to the unStake: multiple checks are the same in both, the values set in the registrationData are the same. Could be refactored maybe in a future PR

**Review by iulianpascalau — APPROVED** [review]

File: vm/systemSmartContracts/staking.go [hunk 4]
```diff
@@ -565,12 +565,15 @@ func (s *stakingSC) unStakeAtEndOfEpoch(args *vmcommon.ContractCallInput) vmcomm
...
+		log.Debug("stakingSC.unStakeAtEndOfEpoch: cannot unStake node which was already unStaked", "blsKey", hex.EncodeToString(args.Arguments[0]))
 		return vmcommon.Ok
 	}
 
-	s.removeFromStakedNodes()
+	if registrationData.Staked {
```
Comment (raduchis) [inline vm/systemSmartContracts/staking.go:573]: could this method call result in too many nodes leaving? the unstakeMethod checks:

```
	if !s.canUnStake() {
		s.eei.AddReturnMessage("unStake is not possible as too many left")
		return vmcommon.UserError
	}
```

File: vm/systemSmartContracts/staking.go [hunk 5]
```diff
@@ -565,12 +565,15 @@ func (s *stakingSC) unStakeAtEndOfEpoch(args *vmcommon.ContractCallInput) vmcomm
...
+		s.removeFromStakedNodes()
+	}
+
 	registrationData.Staked = false
 	registrationData.UnStakedEpoch = s.eei.BlockChainHook().CurrentEpoch()
```
Comment (raduchis) [inline vm/systemSmartContracts/staking.go:585]: if the node was in the waitingList, setting the unstakeEpocha dn UnStakeNonce, means now that he has unbondPeriod?

**Review by raduchis — APPROVED** [review]

**Review by dorin-iancu — APPROVED** [review]

**Review by SebastianMarian — APPROVED** [review]

File: vm/systemSmartContracts/staking.go (hunk 3 above)
Comment (sasurobert) [reply vm/systemSmartContracts/staking.go:568]: yes. it has. first we should remove the flags. after that see what is common on the new version

File: vm/systemSmartContracts/staking.go (hunk 4 above)
Comment (sasurobert) [reply vm/systemSmartContracts/staking.go:573]: here you unStake all - but if you reach this point where cannot unstake - the system will toggle pause.

File: vm/systemSmartContracts/staking.go (hunk 5 above)
Comment (sasurobert) [reply vm/systemSmartContracts/staking.go:585]: yes. it is. but in force unstake cases - it is better to keep the unbonding period.

**Pull request merged**